src/rc-socket.js

```diff
@@ -30,293 +21,341 @@ var root = this;
+
+  // Fix error where close is explicitly called but onopen event is still
+  // triggered.
+  if (this.wasForced) {
+    return this.close();
   }
 
-  // Reconnect
-  setTimeout(this.connect.bind(this), this._getInterval());
+  this.hasOpened = true;
+  this.attempts = 1;
+  this._stateChanged('OPEN', 'onopen', evt);
+  this._sendQueued();
 };
 
-
 /**
- * Loop over all queued messages and send.
- *
  * @private
+ * @memberof RcSocket
+ *
+ * @desc Responsible for interpretting the various possible close types (force,
+ *   retry, refresh, etc...) and reconnecting/proxying events accordinly.
+ *
+ * @param {Object} evt - WebSocket onclose evt.
  */
-RcSocket.prototype._sendQueued = function () {
-  var l = this.queue.length;
-      i = l;
+RcSocket.prototype._onclose = function (evt) {
+  clearTimeout(this.connectTimer);
+  this.ws = null;
+
+  // Because RcSocket holds state we can pass additional information to
+  // upstream handlers regarding why the socket was closed.
+  evt.forced = this.wasForced;
+  evt.isRetrying = this.isRetrying;
+  evt.isRefreshing = this.isRefreshing;
+
+  // Immediately change state and exit on force close.
+  if (this.wasForced) {
+    this._stateChanged('CLOSED', 'onclose', evt);
+
+  // Hack P2: Safegaurd against firefox behavior where close event is
+  // triggered on page navigation and results in an attempted reconnect.
+  } else if (!this.hasUnloaded) {
+    // Was open at some point so we need to trigger close evts
+    // TODO: Wondering it state change should ALWAYS BE CALLED?
+    if (this.hasOpened) {
+      this._trigger('onclose', evt);
+    }
 
-  while (i--) {
-    this._delayQueueSend(i, l - i);
+    this.isRetrying = false;
+    this.isRefreshing = false;
+    this.hasOpened = false;
+    this._reconnect();
   }
 };
 
-
 /**
- * Send delayed message to avoid timing
- * issues when sending queued.
+ * @private
+ * @memberof RcSocket
+ *
+ * @desc Simple proxy for onmessage event.
  *
+ * @param {Object} evt - WebSocket onmessage evt.
+ */
+RcSocket.prototype._onmessage = function (evt) {
+  this._trigger('onmessage', evt);
+};
+
+/**
  * @private
+ * @memberof RcSocket
  *
- * @param - Id/order of message to send.
+ * @desc Simple proxy for onerror event.
+ *
+ * @param {Object} evt - WebSocket onerror evt.
  */
-RcSocket.prototype._delayQueueSend = function (i, d) {
-  var delay = this.delay * d;
-  
-  setTimeout(function () {
-    this.send(this.queue[i]);
-    this.queue.pop();
-  }.bind(this), delay);
+RcSocket.prototype._onerror = function (evt) {
+  this._trigger('onerror', evt);
 };
 
 /**
- * Update state, log, trigger.
+ * @private
+ * @memberof RcSocket
  *
+ * @desc Helper around ws.close to ensure ws exists. If it does not exist we
+ *   fail silently. This seemed logical as closing the socket would have the
+ *   same effect as if the socket never existed. In other words no matter what
+ *   happens in this method the net effect will always be the same.
+ */
```

Comment:
What would happen when we have a timeout firing, to retry a message.  But the socket was forcibly closed.  My understanding is that since ws is undefined, it should not send any message and kill the timeout.  It should not try to open a new socket.

This case would likely not happen.